# CIV1499 - Introduction to Data Science
## Project 2 - Toronto Bike Share

## Setup Notebook

First step is to import all 3rd party libraries required for this analysis

In [1]:
# Importing 3rd party libraries
import os
import pandas as pd
import seaborn as sns
import matplotlib.pylab as plt
import numpy as np
import datetime


# Configuring Notebook
%matplotlib inline
plt.style.use('fivethirtyeight')
sns.set_context("notebook")
import warnings
warnings.filterwarnings('ignore')

# 1. Weather Data

In [2]:
# Set weather data path
#path_weather = os.path.join(os.path.abspath(os.getcwd()), 'data', 'weather')
path_weather = "C:\\Users\\akmar\\Desktop\\Porject 2\\project_2_bikeshare_toronto\\data\\weather"

# Get weather file names
weather_filenames = [filename for filename in os.listdir(path_weather) if 'climate' in filename]

# Print file names
print(weather_filenames[0:5])

['en_climate_hourly_ON_6158359_01-2017_P1H.csv', 'en_climate_hourly_ON_6158359_01-2018_P1H.csv', 'en_climate_hourly_ON_6158359_01-2019_P1H.csv', 'en_climate_hourly_ON_6158359_01-2020_P1H.csv', 'en_climate_hourly_ON_6158359_01-2021_P1H.csv']


# 2. Bike Share Trip Data

In [3]:
# Set bikeshare data path
#path_ridership = os.path.join(os.path.abspath(os.getcwd()), 'data', 'ridership')
path_ridership = "C:\\Users\\akmar\\Desktop\\Porject 2\\project_2_bikeshare_toronto\\data\\ridership"

# Create a list of file names for bike share trips data
trips_filenames = [filename for filename in os.listdir(os.path.join(path_ridership, 'release')) if 'bike_share' in filename]

# Create a dictionary where key:value pairs correspond to the file name and DataFrame respectively
trips_data = {filename: pd.read_csv(os.path.join(path_ridership, 'release', filename)) for filename in trips_filenames}  

# Now lets print out the column names for the first month of each year
print('2017\n{}\n'.format(trips_data['bike_share_2017-1.csv'].columns.tolist()))
print('2018\n{}\n'.format(trips_data['bike_share_2018-1.csv'].columns.tolist()))
print('2019\n{}\n'.format(trips_data['bike_share_2019-1.csv'].columns.tolist()))
print('2020\n{}\n'.format(trips_data['bike_share_2020-1.csv'].columns.tolist()))
print('2021\n{}\n'.format(trips_data['bike_share_2021-1.csv'].columns.tolist()))

2017
['trip_id', 'trip_start_time', 'trip_stop_time', 'trip_duration_seconds', 'from_station_id', 'from_station_name', 'to_station_id', 'to_station_name', 'user_type']

2018
['trip_id', 'trip_duration_seconds', 'from_station_id', 'trip_start_time', 'from_station_name', 'trip_stop_time', 'to_station_id', 'to_station_name', 'user_type']

2019
['Trip Id', 'Subscription Id', 'Trip  Duration', 'Start Station Id', 'Start Time', 'Start Station Name', 'End Station Id', 'End Time', 'End Station Name', 'Bike Id', 'User Type']

2020
['Trip Id', 'Subscription Id', 'Trip  Duration', 'Start Station Id', 'Start Time', 'Start Station Name', 'End Station Id', 'End Time', 'End Station Name', 'Bike Id', 'User Type']

2021
['Trip Id', 'Trip  Duration', 'Start Station Id', 'Start Time', 'Start Station Name', 'End Station Id', 'End Time', 'End Station Name', 'Bike Id', 'User Type']



# 3. Station Data

In [76]:
# Set stations data path
#path_station = os.path.join(os.path.abspath(os.getcwd()), 'data', 'stations')
path_station = "C:\\Users\\akmar\\Desktop\\Porject 2\\project_2_bikeshare_toronto\\data\\stations"

# Import stations
stations = pd.read_csv(os.path.join(path_station, 'bikeshare_stations.csv'))

# View DataFrame
stations.head()

station_id                          name        lat        lon
0        7000  Fort York  Blvd / Capreol Ct  43.639832 -79.395954
1        7001     Wellesley Station Green P  43.664964 -79.383550
2        7002    St. George St / Bloor St W  43.667333 -79.399429
3        7003      Madison Ave / Bloor St W  43.667158 -79.402761
4        7004       University Ave / Elm St  43.656518 -79.389099

# 4. First Step -  Concatenation of CSV files

## 4.1  Weather Data
First, let's take a look at the column names for the first month of each year

In [77]:
# Now lets print out the column names for the first month of each year
print('2017\n{}\n'.format(pd.read_csv(path_weather+"//"+weather_filenames[0],low_memory=False).columns.to_list()))
print('2018\n{}\n'.format(pd.read_csv(path_weather+"//"+weather_filenames[1],low_memory=False).columns.to_list()))
print('2019\n{}\n'.format(pd.read_csv(path_weather+"//"+weather_filenames[2],low_memory=False).columns.to_list()))
print('2020\n{}\n'.format(pd.read_csv(path_weather+"//"+weather_filenames[3],low_memory=False).columns.to_list()))
print('2021\n{}\n'.format(pd.read_csv(path_weather+"//"+weather_filenames[4],low_memory=False).columns.to_list()))


2017
['Longitude (x)', 'Latitude (y)', 'Station Name', 'Climate ID', 'Date/Time', 'Year', 'Month', 'Day', 'Time', 'Temp (°C)', 'Temp Flag', 'Dew Point Temp (°C)', 'Dew Point Temp Flag', 'Rel Hum (%)', 'Rel Hum Flag', 'Wind Dir (10s deg)', 'Wind Dir Flag', 'Wind Spd (km/h)', 'Wind Spd Flag', 'Visibility (km)', 'Visibility Flag', 'Stn Press (kPa)', 'Stn Press Flag', 'Hmdx', 'Hmdx Flag', 'Wind Chill', 'Wind Chill Flag', 'Weather']

2018
['Longitude (x)', 'Latitude (y)', 'Station Name', 'Climate ID', 'Date/Time', 'Year', 'Month', 'Day', 'Time', 'Temp (°C)', 'Temp Flag', 'Dew Point Temp (°C)', 'Dew Point Temp Flag', 'Rel Hum (%)', 'Rel Hum Flag', 'Wind Dir (10s deg)', 'Wind Dir Flag', 'Wind Spd (km/h)', 'Wind Spd Flag', 'Visibility (km)', 'Visibility Flag', 'Stn Press (kPa)', 'Stn Press Flag', 'Hmdx', 'Hmdx Flag', 'Wind Chill', 'Wind Chill Flag', 'Weather']

2019
['Longitude (x)', 'Latitude (y)', 'Station Name', 'Climate ID', 'Date/Time', 'Year', 'Month', 'Day', 'Time', 'Temp (°C)', 'Temp F

As it can be seen, there are issues with the consistency of names of columns like `Date/Time` and `Time`. Also, `2021 dataset` has two additional columns including, `Precip. Amount (mm)`, `Precip. Amount Flag` which are not reported in the other datasets. So, in the following line, we try to concatenate these weather csv files based on the corrected column names, as follows:
- Define a dictionary which its key are the unique column names for the concatenated dataset and the values are column names for each key in different datasets
- Use the dictionary to change the name of each CSV file and 
- then, concatenate those CSV files


In [78]:
#Creating a Dictionary for common columns' names in weather datasets
weather_col_dic ={
    "Date/Time": ["Date/Time", "Date/Time (LST)"],
    "Time": ["Time", "Time (LST)"],
}

#Create a new DF for Weather
weather_df = pd.DataFrame()
for i in range(len(weather_filenames)):
    temp_df = pd.read_csv(path_weather+"//"+weather_filenames[i],low_memory=False)
    for col in temp_df.columns:
        for key,val in weather_col_dic.items():

            for item in val:
                if col == item:
                    temp_df = temp_df.rename(columns={col:key})
    
    
    temp_df['Excel_File'] = str(weather_filenames[i].split('_')[5])
    weather_df = pd.concat([weather_df,temp_df],
                           ignore_index=True, sort = False, axis=0)

    
#Weather Dataframe
weather_df.head()

Longitude (x)  Latitude (y)         Station Name  Climate ID  \
0          -79.4         43.63  TORONTO CITY CENTRE     6158359   
1          -79.4         43.63  TORONTO CITY CENTRE     6158359   
2          -79.4         43.63  TORONTO CITY CENTRE     6158359   
3          -79.4         43.63  TORONTO CITY CENTRE     6158359   
4          -79.4         43.63  TORONTO CITY CENTRE     6158359   

          Date/Time  Year  Month  Day   Time  Temp (°C)  ... Stn Press (kPa)  \
0  2017-01-01 00:00  2017      1    1  00:00        1.5  ...           99.81   
1  2017-01-01 01:00  2017      1    1  01:00        1.5  ...          100.01   
2  2017-01-01 02:00  2017      1    1  02:00        1.0  ...          100.14   
3  2017-01-01 03:00  2017      1    1  03:00        1.2  ...          100.32   
4  2017-01-01 04:00  2017      1    1  04:00        1.3  ...          100.48   

   Stn Press Flag Hmdx  Hmdx Flag Wind Chill  Wind Chill Flag Weather  \
0             NaN  NaN        NaN        NaN              NaN     NaN   
1             NaN  NaN        NaN        NaN              NaN     NaN   
2             NaN  NaN        NaN        NaN              NaN     NaN   
3             NaN  NaN        NaN        NaN              NaN     NaN   
4             NaN  NaN        NaN        NaN              NaN     NaN   

   Excel_File Precip. Amount (mm)  Precip. Amount Flag  
0     01-2017                 NaN                  NaN  
1     01-2017                 NaN                  NaN  
2     01-2017                 NaN                  NaN  
3     01-2017                 NaN                  NaN  
4     01-2017                 NaN                  NaN  

[5 rows x 31 columns]

Among all these columns, we are going to use only some of them. Then, it is better to drop some of them. After checking the dataset, it was decided to drop the following columns from the `Weather_df` dataset:
`["Station Name", "Climate ID", "Temp Flag", "Dew Point Temp Flag", "Rel Hum Flag", "Wind Dir Flag", "Wind Spd Flag", "Visibility Flag", "Hmdx Flag", "Wind Chill Flag", "Stn Press Flag", "Excel_File", "Precip. Amount (mm)", "Precip. Amount Flag"]`

In [79]:
lst_weather_drop_col = ["Station Name", "Climate ID", "Temp Flag", "Dew Point Temp Flag", "Rel Hum Flag", "Wind Dir Flag",
                        "Wind Spd Flag", "Visibility Flag", "Hmdx Flag", "Wind Chill Flag","Stn Press Flag", "Excel_File", "Precip. Amount (mm)",
                        "Precip. Amount Flag"]
weather_df = weather_df.drop(lst_weather_drop_col,axis = 1)
weather_df.head()

Longitude (x)  Latitude (y)         Date/Time  Year  Month  Day   Time  \
0          -79.4         43.63  2017-01-01 00:00  2017      1    1  00:00   
1          -79.4         43.63  2017-01-01 01:00  2017      1    1  01:00   
2          -79.4         43.63  2017-01-01 02:00  2017      1    1  02:00   
3          -79.4         43.63  2017-01-01 03:00  2017      1    1  03:00   
4          -79.4         43.63  2017-01-01 04:00  2017      1    1  04:00   

   Temp (°C)  Dew Point Temp (°C)  Rel Hum (%)  Wind Dir (10s deg)  \
0        1.5                 -3.6         69.0                26.0   
1        1.5                 -3.9         67.0                27.0   
2        1.0                 -4.3         68.0                26.0   
3        1.2                 -4.3         67.0                26.0   
4        1.3                 -4.4         66.0                26.0   

   Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)  Hmdx  Wind Chill Weather  
0             39.0             16.1            99.81   NaN         NaN     NaN  
1             35.0             16.1           100.01   NaN         NaN     NaN  
2             32.0             16.1           100.14   NaN         NaN     NaN  
3             37.0             16.1           100.32   NaN         NaN     NaN  
4             28.0             16.1           100.48   NaN         NaN     NaN

We should convert the `Date/Time` column into a `datetime format` and localize the datetime to `EST`. In order to do this, we would use the following codes

In [80]:
weather_df['Date/Time'] = pd.to_datetime(weather_df['Date/Time'])
weather_df['Date/Time'] = weather_df.apply(lambda row: row['Date/Time'].tz_localize(tz ='EST'), axis = 1)

Let's take a look at the weather dataframe.

In [81]:
weather_df.isnull().sum()

Longitude (x)              0
Latitude (y)               0
Date/Time                  0
Year                       0
Month                      0
Day                        0
Time                       0
Temp (°C)                149
Dew Point Temp (°C)      224
Rel Hum (%)              211
Wind Dir (10s deg)      2531
Wind Spd (km/h)           69
Visibility (km)           71
Stn Press (kPa)          152
Hmdx                   36704
Wind Chill             36253
Weather                37057
dtype: int64

There are 37,057 null values for the Weather column. It is assumed that null values represent a 'Clear' weather type. Let's fill thees null values with Clear

In [82]:
# Fill NaN values with "Clear" 
weather_df['Weather'] = weather_df['Weather'].fillna('Clear')#(method='ffill')

In [83]:
weather_df.isnull().sum()

Longitude (x)              0
Latitude (y)               0
Date/Time                  0
Year                       0
Month                      0
Day                        0
Time                       0
Temp (°C)                149
Dew Point Temp (°C)      224
Rel Hum (%)              211
Wind Dir (10s deg)      2531
Wind Spd (km/h)           69
Visibility (km)           71
Stn Press (kPa)          152
Hmdx                   36704
Wind Chill             36253
Weather                    0
dtype: int64

## 4.2  Bike Share Trip Data


Same to the weather datasets, there are some issues with the consistency of names of columns and the number of reported columns in each dataset. So, in the following line, we try to concatenate these Bikeride share csv files based on the corrected column names, as follows:
- Define a dictionary which its key are the unique column names for the concatenated dataset and the values are column names for each key in different datasets
- Use the dictionary to change the name of each CSV file and 
- then, concatenate those CSV files

In [11]:
#Creating a Dictionary of columns' name
bike_col_dic ={
    "ID": ["trip_id", "Trip Id", "Trip Id"],
    "Start_Time": ["trip_start_time", "Start Time"],
    "End_Time": ["trip_stop_time","End Time"],
    "Duration": ["trip_duration_seconds","Trip  Duration"],
    "Start_Station_id": ["from_station_id","Start Station Id"],
    "Start_Station_name": ["from_station_name","Start Station Name"],
    "End_Station_id": ["to_station_id","End Station Id"],
    "End_Station_name": ["to_station_name","End Station Name"],
    "User_Type": ["user_type","User Type"]
}
#bike_col_dic


#Create a new DF for BikeRides
bikerides_df = pd.DataFrame()


for i in range(len(trips_filenames)):
    temp_df = pd.read_csv(path_ridership+"\\release\\"+trips_filenames[i], low_memory=False)
    for col in temp_df.columns:
        for key,val in bike_col_dic.items():
            
            for item in val:
                if col == item:
                    temp_df = temp_df.rename(columns={col:key})

    bikerides_df = pd.concat([bikerides_df,temp_df],
                             ignore_index=True, sort = False, axis=0)

# View DataFrame
bikerides_df.head()

ID              Start_Time                End_Time  Duration  \
0  712441  2017-01-01 05:03 (UTC)  2017-01-01 05:08 (UTC)       274   
1  712442  2017-01-01 05:03 (UTC)  2017-01-01 05:12 (UTC)       538   
2  712443  2017-01-01 05:05 (UTC)  2017-01-01 05:22 (UTC)       992   
3  712444  2017-01-01 05:09 (UTC)  2017-01-01 05:26 (UTC)      1005   
4  712445  2017-01-01 05:14 (UTC)  2017-01-01 05:25 (UTC)       645   

   Start_Station_id                         Start_Station_name  \
0            7006.0            Bay St / College St (East Side)   
1            7046.0                 Niagara St / Richmond St W   
2            7048.0  Front St / Yonge St (Hockey Hall of Fame)   
3            7177.0            East Liberty St / Pirandello St   
4            7203.0                Bathurst St / Queens Quay W   

   End_Station_id                  End_Station_name User_Type  \
0          7021.0                Bay St / Albert St    Member   
1          7147.0            King St W / Fraser Ave    Member   
2          7089.0              Church St  / Wood St    Member   
3          7202.0  Queen St W / York St (City Hall)    Member   
4          7010.0           King St W / Spadina Ave    Member   

   Subscription Id  Bike Id  
0              NaN      NaN  
1              NaN      NaN  
2              NaN      NaN  
3              NaN      NaN  
4              NaN      NaN

After concatenation of different CSV files, the next step is to convert `Start_time` and `End_time` into `datetime` format. To do this, first we should take a look at these two columns. As it can be seen, these two columns contains string values such as `(UTC)` and `(ETS)`. So, we should remove this string from these two columns, as follow:
- add a new column containing the time zone of `Strat_Time` column. we use this column later to localize these two columns since they are localized based on different time zones
- removing values `(UTC)` and `(ETS)` from both colums

In [12]:
bikerides_df['Time_Zone'] = bikerides_df.apply(
    lambda row: row["Start_Time"].split("(")[1].split(")")[0] if ("(" in row["Start_Time"]) else False, axis=1)     
bikerides_df['Start_Time'] = bikerides_df.apply(lambda row: row["Start_Time"].split("(")[0], axis=1)
bikerides_df['End_Time']   = bikerides_df.apply(
    lambda row: row["End_Time"].split("(")[0] if type(row["End_Time"]) == str else "delete", axis=1)

In [13]:
bikerides_df.head()

ID         Start_Time           End_Time  Duration  Start_Station_id  \
0  712441  2017-01-01 05:03   2017-01-01 05:08        274            7006.0   
1  712442  2017-01-01 05:03   2017-01-01 05:12        538            7046.0   
2  712443  2017-01-01 05:05   2017-01-01 05:22        992            7048.0   
3  712444  2017-01-01 05:09   2017-01-01 05:26       1005            7177.0   
4  712445  2017-01-01 05:14   2017-01-01 05:25        645            7203.0   

                          Start_Station_name  End_Station_id  \
0            Bay St / College St (East Side)          7021.0   
1                 Niagara St / Richmond St W          7147.0   
2  Front St / Yonge St (Hockey Hall of Fame)          7089.0   
3            East Liberty St / Pirandello St          7202.0   
4                Bathurst St / Queens Quay W          7010.0   

                   End_Station_name User_Type  Subscription Id  Bike Id  \
0                Bay St / Albert St    Member              NaN      NaN   
1            King St W / Fraser Ave    Member              NaN      NaN   
2              Church St  / Wood St    Member              NaN      NaN   
3  Queen St W / York St (City Hall)    Member              NaN      NaN   
4           King St W / Spadina Ave    Member              NaN      NaN   

  Time_Zone  
0       UTC  
1       UTC  
2       UTC  
3       UTC  
4       UTC

There are some error in converting `End_Time` column into a datetime format which arises from that some values in this column are either nan or null. So, we should find these values and drop their records from the dataset

In [14]:
bikerides_df = bikerides_df.drop(index = bikerides_df[bikerides_df["End_Time"] == "delete"].index[0])

After removing the string parts from these two columns, we simply convert them to `datetime` format using following line

In [15]:
bikerides_df['Start_Time'] = pd.to_datetime(bikerides_df['Start_Time'])
bikerides_df['End_Time'] = pd.to_datetime(bikerides_df['End_Time'])  

Now, `Start_time` and `End_time` are of the datetime form, but they strill contain two different time zones. so in order to make it possible to work with them, we should convert all time zone into `EST`. To do this, first we define a function, then apply the function on each column to have uniform timezones

In [16]:
def UTC_To_EST(x):
    """
    Converting datetime for UTC to EST
    """
    UTC_TO_EST = pd.Timestamp(x).tz_localize('UTC')
    return UTC_TO_EST.tz_convert('EST')


bikerides_df['Start_Time'] = bikerides_df.apply(
    lambda row: UTC_To_EST(row['Start_Time']) if row["Time_Zone"] == "UTC" else row['Start_Time'], axis=1)
bikerides_df['End_Time'] = bikerides_df.apply(
    lambda row: UTC_To_EST(row['End_Time']) if row["Time_Zone"] == "UTC" else row['End_Time'], axis=1)

In [17]:
bikerides_df.head()

ID                 Start_Time                   End_Time  Duration  \
0  712441  2017-01-01 00:03:00-05:00  2017-01-01 00:08:00-05:00       274   
1  712442  2017-01-01 00:03:00-05:00  2017-01-01 00:12:00-05:00       538   
2  712443  2017-01-01 00:05:00-05:00  2017-01-01 00:22:00-05:00       992   
3  712444  2017-01-01 00:09:00-05:00  2017-01-01 00:26:00-05:00      1005   
4  712445  2017-01-01 00:14:00-05:00  2017-01-01 00:25:00-05:00       645   

   Start_Station_id                         Start_Station_name  \
0            7006.0            Bay St / College St (East Side)   
1            7046.0                 Niagara St / Richmond St W   
2            7048.0  Front St / Yonge St (Hockey Hall of Fame)   
3            7177.0            East Liberty St / Pirandello St   
4            7203.0                Bathurst St / Queens Quay W   

   End_Station_id                  End_Station_name User_Type  \
0          7021.0                Bay St / Albert St    Member   
1          7147.0            King St W / Fraser Ave    Member   
2          7089.0              Church St  / Wood St    Member   
3          7202.0  Queen St W / York St (City Hall)    Member   
4          7010.0           King St W / Spadina Ave    Member   

   Subscription Id  Bike Id Time_Zone  
0              NaN      NaN       UTC  
1              NaN      NaN       UTC  
2              NaN      NaN       UTC  
3              NaN      NaN       UTC  
4              NaN      NaN       UTC

In this step, we try to localize all datetime to EST since, in the previous step, we only convert `UTC` ones to `EST`

In [18]:
#Localizing Timestamps 
bikerides_df["Start_Time"] = bikerides_df.apply(
    lambda row: row["Start_Time"] if (
        str(row["Start_Time"].tzinfo) == "EST") else row["Start_Time"].tz_localize(tz ='EST'), axis = 1)
bikerides_df["End_Time"] = bikerides_df.apply(
    lambda row: row["End_Time"] if (
        str(row["End_Time"].tzinfo) == "EST") else row["End_Time"].tz_localize(tz ='EST'), axis = 1) 

In [19]:
bikerides_df.head()

ID                Start_Time                  End_Time  Duration  \
0  712441 2017-01-01 00:03:00-05:00 2017-01-01 00:08:00-05:00       274   
1  712442 2017-01-01 00:03:00-05:00 2017-01-01 00:12:00-05:00       538   
2  712443 2017-01-01 00:05:00-05:00 2017-01-01 00:22:00-05:00       992   
3  712444 2017-01-01 00:09:00-05:00 2017-01-01 00:26:00-05:00      1005   
4  712445 2017-01-01 00:14:00-05:00 2017-01-01 00:25:00-05:00       645   

   Start_Station_id                         Start_Station_name  \
0            7006.0            Bay St / College St (East Side)   
1            7046.0                 Niagara St / Richmond St W   
2            7048.0  Front St / Yonge St (Hockey Hall of Fame)   
3            7177.0            East Liberty St / Pirandello St   
4            7203.0                Bathurst St / Queens Quay W   

   End_Station_id                  End_Station_name User_Type  \
0          7021.0                Bay St / Albert St    Member   
1          7147.0            King St W / Fraser Ave    Member   
2          7089.0              Church St  / Wood St    Member   
3          7202.0  Queen St W / York St (City Hall)    Member   
4          7010.0           King St W / Spadina Ave    Member   

   Subscription Id  Bike Id Time_Zone  
0              NaN      NaN       UTC  
1              NaN      NaN       UTC  
2              NaN      NaN       UTC  
3              NaN      NaN       UTC  
4              NaN      NaN       UTC

Let's take a look at the `Duration`. We should check that if there is any missing.

In [20]:
bikerides_df["Duration"].isnull().sum()

0

We can see that there is no missing values for `Duration`. Lets check that all values are of integer type or not.

Lets take a look at `User_Type`.

we can see that we have different terms for each of annual and casual members. As it can be seen we have two keywords for Annual Members and two for casual members

Lets make it uniform

In [21]:
bikerides_df["User_Type"].unique()

array(['Member', 'Casual', 'Annual Member', 'Casual Member'], dtype=object)

In [22]:
bikerides_df["User_Type"] = bikerides_df["User_Type"].apply(lambda row: "Casual" if ((row == "Casual") | (row == "Casual Member")) else "Annual")

Let's check the unique values of `User_Type` columns. 

There we go

In [23]:
bikerides_df["User_Type"].unique()

array(['Annual', 'Casual'], dtype=object)

We don't need `Time_zone` column anymore. so let's drop it.

In [24]:
bikerides_df.drop("Time_Zone", axis = 1, inplace = True)

Let's take a another look at the dataframe

In [25]:
bikerides_df.head()

ID                Start_Time                  End_Time  Duration  \
0  712441 2017-01-01 00:03:00-05:00 2017-01-01 00:08:00-05:00       274   
1  712442 2017-01-01 00:03:00-05:00 2017-01-01 00:12:00-05:00       538   
2  712443 2017-01-01 00:05:00-05:00 2017-01-01 00:22:00-05:00       992   
3  712444 2017-01-01 00:09:00-05:00 2017-01-01 00:26:00-05:00      1005   
4  712445 2017-01-01 00:14:00-05:00 2017-01-01 00:25:00-05:00       645   

   Start_Station_id                         Start_Station_name  \
0            7006.0            Bay St / College St (East Side)   
1            7046.0                 Niagara St / Richmond St W   
2            7048.0  Front St / Yonge St (Hockey Hall of Fame)   
3            7177.0            East Liberty St / Pirandello St   
4            7203.0                Bathurst St / Queens Quay W   

   End_Station_id                  End_Station_name User_Type  \
0          7021.0                Bay St / Albert St    Annual   
1          7147.0            King St W / Fraser Ave    Annual   
2          7089.0              Church St  / Wood St    Annual   
3          7202.0  Queen St W / York St (City Hall)    Annual   
4          7010.0           King St W / Spadina Ave    Annual   

   Subscription Id  Bike Id  
0              NaN      NaN  
1              NaN      NaN  
2              NaN      NaN  
3              NaN      NaN  
4              NaN      NaN

we should check `Subscription Id` and `Bike Id` columns. 

In [26]:
bikerides_df["Subscription Id"].describe().dtype

dtype('float64')

In [27]:
bikerides_df["Bike Id"].describe().dtype

dtype('float64')

As it can be seen, these two columns are of the float type. let's convert them into an integer format

In [28]:
bikerides_df["Subscription Id"] = bikerides_df["Subscription Id"].astype('Int64')
bikerides_df["Bike Id"] = bikerides_df["Bike Id"].astype('Int64')
bikerides_df.head()

ID                Start_Time                  End_Time  Duration  \
0  712441 2017-01-01 00:03:00-05:00 2017-01-01 00:08:00-05:00       274   
1  712442 2017-01-01 00:03:00-05:00 2017-01-01 00:12:00-05:00       538   
2  712443 2017-01-01 00:05:00-05:00 2017-01-01 00:22:00-05:00       992   
3  712444 2017-01-01 00:09:00-05:00 2017-01-01 00:26:00-05:00      1005   
4  712445 2017-01-01 00:14:00-05:00 2017-01-01 00:25:00-05:00       645   

   Start_Station_id                         Start_Station_name  \
0            7006.0            Bay St / College St (East Side)   
1            7046.0                 Niagara St / Richmond St W   
2            7048.0  Front St / Yonge St (Hockey Hall of Fame)   
3            7177.0            East Liberty St / Pirandello St   
4            7203.0                Bathurst St / Queens Quay W   

   End_Station_id                  End_Station_name User_Type  \
0          7021.0                Bay St / Albert St    Annual   
1          7147.0            King St W / Fraser Ave    Annual   
2          7089.0              Church St  / Wood St    Annual   
3          7202.0  Queen St W / York St (City Hall)    Annual   
4          7010.0           King St W / Spadina Ave    Annual   

   Subscription Id  Bike Id  
0             <NA>     <NA>  
1             <NA>     <NA>  
2             <NA>     <NA>  
3             <NA>     <NA>  
4             <NA>     <NA>

let's take a look at `Start_Station_id`, `Start_Station_name`, `End_Station_id`, and `End_Station_name`. There are two problems in these four columns.
- First, out of 12 Million records, more than one million records have no either `Start_Station_id` or `End_Station_id` or both.
- Second, There are issues with the consistency of station names in both `Start_Station_name` and `End_Station_name`. 

To solve these problems, firstly, we try to convert `Start_Station_id` and `End_Station_id` from `float` type to `integer` format. Then we will try to fill missing cells of `Start_Station_id` and `End_Station_id` with correct values. 
Let's convert these columns to integer.


In [29]:
bikerides_df["Start_Station_id"] = bikerides_df["Start_Station_id"].astype('Int64')
bikerides_df["End_Station_id"] = bikerides_df["End_Station_id"].astype('Int64')
bikerides_df.head()


ID                Start_Time                  End_Time  Duration  \
0  712441 2017-01-01 00:03:00-05:00 2017-01-01 00:08:00-05:00       274   
1  712442 2017-01-01 00:03:00-05:00 2017-01-01 00:12:00-05:00       538   
2  712443 2017-01-01 00:05:00-05:00 2017-01-01 00:22:00-05:00       992   
3  712444 2017-01-01 00:09:00-05:00 2017-01-01 00:26:00-05:00      1005   
4  712445 2017-01-01 00:14:00-05:00 2017-01-01 00:25:00-05:00       645   

   Start_Station_id                         Start_Station_name  \
0              7006            Bay St / College St (East Side)   
1              7046                 Niagara St / Richmond St W   
2              7048  Front St / Yonge St (Hockey Hall of Fame)   
3              7177            East Liberty St / Pirandello St   
4              7203                Bathurst St / Queens Quay W   

   End_Station_id                  End_Station_name User_Type  \
0            7021                Bay St / Albert St    Annual   
1            7147            King St W / Fraser Ave    Annual   
2            7089              Church St  / Wood St    Annual   
3            7202  Queen St W / York St (City Hall)    Annual   
4            7010           King St W / Spadina Ave    Annual   

   Subscription Id  Bike Id  
0             <NA>     <NA>  
1             <NA>     <NA>  
2             <NA>     <NA>  
3             <NA>     <NA>  
4             <NA>     <NA>

Afterward we will fill missing values. A good way to fill this missing values is to use the `station names` columns. Firstly, we will create a dictionary of station names and their ids, and then we will apply this dictionary to the missing values.

In [30]:
#Creating two temporary dataframe for extracting unique station names and satation ids
df_temp1_station = bikerides_df[["Start_Station_name","Start_Station_id","Duration"]].groupby([
    "Start_Station_name","Start_Station_id"]).count().sort_values(by="Start_Station_id").reset_index().drop(
    "Duration", axis =1).rename(columns={"Start_Station_name": "name", "Start_Station_id": "id"})

df_temp2_station =bikerides_df[["End_Station_name","End_Station_id","Duration"]].groupby([
    "End_Station_name","End_Station_id"]).count().sort_values(by="End_Station_id").reset_index().drop(
    "Duration", axis =1).rename(columns={"End_Station_name": "name", "End_Station_id": "id"})

#Concatenation of two temporary dataframe to have a dataframe of unique names and ids
df_temp_station = pd.concat([df_temp1_station, df_temp2_station], ignore_index=True).sort_values(
    by="id").drop_duplicates(subset=["name","id"], keep='first')

# Converting dataframe to a dictionary
dicts_temp = df_temp_station.groupby('id')['name'].agg(list).to_dict()
dicts_temp_inv = df_temp_station.groupby('name')['id'].agg(list).to_dict()

dicts_temp[8000] = ["Michael Sweet Ave / St. Patrick St"]
dicts_temp_inv["Michael Sweet Ave / St. Patrick St"] = [8000]

In [31]:
#finding the indices of records with either no Start_Staion_id or End_Staion_id
temp_lst_no_start_id = bikerides_df[bikerides_df["Start_Station_id"].isnull() == True].index.tolist()
temp_lst_no_end_id = bikerides_df[bikerides_df["End_Station_id"].isnull() == True].index.tolist()

# temp_lst_no_start_id = temp_lst_no_start_id[0:4]

In [32]:
#filling missing values in Start_station_id column
for items in temp_lst_no_start_id:
    if bikerides_df["Start_Station_name"].iloc[items] in dicts_temp_inv.keys():

        bikerides_df["Start_Station_id"].iloc[items] = dicts_temp_inv.get(
            bikerides_df["Start_Station_name"].iloc[items])[0]
    else:

        bikerides_df["Start_Station_id"].iloc[items] = 1000
        

In [33]:
#filling missing values in End_station_id column
for items in temp_lst_no_end_id:
    if bikerides_df["End_Station_name"].iloc[items] in dicts_temp_inv.keys():

        bikerides_df["End_Station_id"].iloc[items] = dicts_temp_inv.get(
            bikerides_df["End_Station_name"].iloc[items])[0]
    else:

        bikerides_df["End_Station_id"].iloc[items] = 2000

Lets take a look at the dataframe 

In [34]:
bikerides_df.head()

ID                Start_Time                  End_Time  Duration  \
0  712441 2017-01-01 00:03:00-05:00 2017-01-01 00:08:00-05:00       274   
1  712442 2017-01-01 00:03:00-05:00 2017-01-01 00:12:00-05:00       538   
2  712443 2017-01-01 00:05:00-05:00 2017-01-01 00:22:00-05:00       992   
3  712444 2017-01-01 00:09:00-05:00 2017-01-01 00:26:00-05:00      1005   
4  712445 2017-01-01 00:14:00-05:00 2017-01-01 00:25:00-05:00       645   

   Start_Station_id                         Start_Station_name  \
0              7006            Bay St / College St (East Side)   
1              7046                 Niagara St / Richmond St W   
2              7048  Front St / Yonge St (Hockey Hall of Fame)   
3              7177            East Liberty St / Pirandello St   
4              7203                Bathurst St / Queens Quay W   

   End_Station_id                  End_Station_name User_Type  \
0            7021                Bay St / Albert St    Annual   
1            7147            King St W / Fraser Ave    Annual   
2            7089              Church St  / Wood St    Annual   
3            7202  Queen St W / York St (City Hall)    Annual   
4            7010           King St W / Spadina Ave    Annual   

   Subscription Id  Bike Id  
0             <NA>     <NA>  
1             <NA>     <NA>  
2             <NA>     <NA>  
3             <NA>     <NA>  
4             <NA>     <NA>

In this stage, we will drop records which we could not find a matching id for their stations. Out of more than 12 Million records, we drop nearly 36000 records which is about 0.3% of data.

In [36]:
bikerides_df = bikerides_df.drop(bikerides_df[
    (bikerides_df["Start_Station_id"] == 1000) | (bikerides_df["End_Station_id"] == 2000) ].index)
bikerides_df.shape[0]

12163646

In [37]:
bikerides_df.describe()

ID      Duration  Start_Station_id  End_Station_id  \
count  1.216365e+07  1.216365e+07      1.216364e+07    1.216007e+07   
mean   7.649343e+06  1.059789e+03      7.203452e+03    7.202321e+03   
std    4.005918e+06  1.003544e+04      1.642270e+02    1.639480e+02   
min    7.123820e+05  0.000000e+00      7.000000e+03    7.000000e+03   
25%    4.182515e+06  4.440000e+02      7.059000e+03    7.059000e+03   
50%    7.632616e+06  7.340000e+02      7.173000e+03    7.172000e+03   
75%    1.111598e+07  1.158000e+03      7.298000e+03    7.298000e+03   
max    1.463520e+07  1.240378e+07      8.000000e+03    8.000000e+03   

       Subscription Id       Bike Id  
count     5.052207e+06  8.779829e+06  
mean      4.852131e+05  3.457701e+03  
std       1.668191e+05  1.851456e+03  
min       6.537700e+04  1.400000e+01  
25%       3.373850e+05  2.035000e+03  
50%       4.752030e+05  3.538000e+03  
75%       6.056665e+05  4.899000e+03  
max       8.634190e+05  6.936000e+03

As seen above, there are outliers under the duration column. These outliers could affect our analysis and it is important to remove them. 

By looking at the decription, the maximum duration of ride is 4 orders of magnitude larger than the 75th percentile. 
The maximum trip duration shown is over 12 million seconds (~140 days). This looks like a significant outlier.
There are also trip duration with 0 seconds that are included in the data. We need to remove them from the data.
Let's use the IQR for the duration column and remove values falling above or under ±1.5*IQR


In [39]:
# Cleaning the dataset
def remove_outliers(df, variable):
    q1 = df[variable].quantile(0.25)
    q3 = df[variable].quantile(0.75)
    iqr = q3 - q1
    lower_limit = q1 - 1.5 * iqr
    upper_limit = q3 + 1.5 * iqr
#     print(upper_limit)
    return df[(df[variable] < upper_limit)]
bikerides_df = remove_outliers(bikerides_df,"Duration")

#Converting From float to Integer
bikerides_df.shape[0]

11661157

We've removed about 502,489 entries. This is about 4% of the total trips data

### Let's review our dataframes

In [43]:
bikerides_df.head(2)

ID                Start_Time                  End_Time  Duration  \
0  712441 2017-01-01 00:03:00-05:00 2017-01-01 00:08:00-05:00       274   
1  712442 2017-01-01 00:03:00-05:00 2017-01-01 00:12:00-05:00       538   

   Start_Station_id               Start_Station_name  End_Station_id  \
0              7006  Bay St / College St (East Side)            7021   
1              7046       Niagara St / Richmond St W            7147   

         End_Station_name User_Type  Subscription Id  Bike Id  
0      Bay St / Albert St    Annual             <NA>     <NA>  
1  King St W / Fraser Ave    Annual             <NA>     <NA>

In [96]:
stations.head(2)

station_id                          name        lat        lon
0        7000  Fort York  Blvd / Capreol Ct  43.639832 -79.395954
1        7001     Wellesley Station Green P  43.664964 -79.383550

In [97]:
weather_df.head(2)

Longitude (x)  Latitude (y)                 Date/Time  Year  Month  Day  \
0          -79.4         43.63 2017-01-01 00:00:00-05:00  2017      1    1   
1          -79.4         43.63 2017-01-01 01:00:00-05:00  2017      1    1   

    Time  Temp (°C)  Dew Point Temp (°C)  Rel Hum (%)  Wind Dir (10s deg)  \
0  00:00        1.5                 -3.6         69.0                26.0   
1  01:00        1.5                 -3.9         67.0                27.0   

   Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)  Hmdx  Wind Chill Weather  
0             39.0             16.1            99.81   NaN         NaN   Clear  
1             35.0             16.1           100.01   NaN         NaN   Clear

### Export all three dataframes into a csv file to be used in part 02 - EDA

In [62]:
#Exporting the bikeridership dataframe
bikerides_df.to_csv('bikerides_df.csv', index=False )

In [84]:
#Exporting the DataFrame
weather_df.to_csv('weather_df.csv', index=False )

In [64]:
#Exporting the station dataframe
stations.to_csv('stations.csv', index=False )

In [87]:
print("done")

done


In [95]:
bikerides_df.tail()

ID                Start_Time                  End_Time  \
12195477  13916477 2021-09-30 23:59:00-05:00 2021-01-10 00:08:00-05:00   
12195478  13916478 2021-09-30 23:59:00-05:00 2021-01-10 00:11:00-05:00   
12195479  13916479 2021-09-30 23:59:00-05:00 2021-01-10 00:22:00-05:00   
12195480  13916480 2021-09-30 23:59:00-05:00 2021-01-10 00:14:00-05:00   
12195481  13916481 2021-09-30 23:59:00-05:00 2021-01-10 00:03:00-05:00   

          Duration  Start_Station_id  \
12195477       553              7274   
12195478       707              7204   
12195479      1347              7372   
12195480       899              7002   
12195481       203              7026   

                                  Start_Station_name  End_Station_id  \
12195477  Queen's Park Cres E / Grosvenor St - SMART            7382   
12195478                    College St / Crawford St            7020   
12195479                     King St W / Portland St            7273   
12195480                  St. George St / Bloor St W            7463   
12195481                      Bay St / St. Joseph St            7044   

                     End_Station_name User_Type  Subscription Id  Bike Id  
12195477    Simcoe St / Adelaide St W    Annual             <NA>     3624  
12195478      Phoebe St / Spadina Ave    Annual             <NA>     6513  
12195479  Bay St / Charles St - SMART    Casual             <NA>     2784  
12195480     Adelaide St W / Brant St    Annual             <NA>     5867  
12195481     Church St / Alexander St    Annual             <NA>     2234